plugin.video.hrmagplayer/libplayer/player.py

```diff
@@ -0,0 +1,134 @@
+    # Add a sort method for the virtual folder items (alphabetically, ignore articles)
+    #xbmcplugin.addSortMethod(__handle__, xbmcplugin.SORT_METHOD_LABEL_IGNORE_THE)
+    # Finish creating a virtual folder.
+    xbmcplugin.endOfDirectory(__handle__)
+    
+def list_episodes(context, show):
+    # Get video categories
+    showid = context['showList']
+    episodes = context['episodes']
+    # Create a list for our items.
+    listing = []
+    # Iterate through categories
+    index = 0
+    for episode in episodes:
+        # Create a list item with a text label and a thumbnail image.
+        list_item = xbmcgui.ListItem(label=episode['title'], thumbnailImage=episode['image'])
+        #list_item.setInfo('video', 'Test Info')
+        url = '{0}?action=play&url={1}&episode={2}&show={3}'.format(__url__, episode['link'], encode(episode['title']), show)
+        is_folder = False
+        listing.append((url, list_item, is_folder))
+        index += 1
+    xbmcplugin.addDirectoryItems(__handle__, listing, len(listing))
+    xbmcplugin.endOfDirectory(__handle__)
+    
+def dispatch(url, handle, parameter):
+    logStr = "Entry: " + url + ", " + str(handle) + ", " + parameter
+    xbmc.log(logStr, xbmc.LOGINFO)
+    parameters = extractParameters(parameter)
+    action = None
+    show = None
+    url = None
+    episode = None
+    if parameters != None:
+        for action in parameters:
+            logStr = action + " = " + parameters[action]
+            xbmc.log(logStr, xbmc.LOGINFO)
+        if 'action' in parameters:
+            action = parameters['action']
+        if 'show' in parameters:
+            show = int(parameters['show'])
+        if 'url' in parameters:
+            url = parameters['url']
+        if 'episode' in parameters:
+            episode = parameters['episode']
+    else:
+        xbmc.log("No actions", xbmc.LOGINFO)
+        action = None   
+        
+    context = ChannelContext(xbmcaddon.Addon())
+
+    if action == None:
+        list_shows(context)
+    elif action == 'listing':
+        loader = ChannelLoader()
+        loader.loadEpisodeList(context, show)
+        # Play live stream immediately
+        if show == 0:
+            video = context['episodes'][0]['link']
+            resolved_video = loader.resolveLiveUrl(context, video)
+            listitem = xbmcgui.ListItem('')
+            listitem.setInfo('video', {'Title': 'Livestream', 'Genre': ''})
+            player = HrMagPlayer()
+            player.play(resolved_video, listitem)
+        else:
+            list_episodes(context, show)
+    elif action == 'play':
+        video = getVideoLink(url)
+        listitem = xbmcgui.ListItem('')
+        id = getShowId(context, show)
+        title = context['showList'][show]['name']
+        ep = decode(episode)
+        #ep = context['episodes'][episode]
+        listitem.setInfo('video', {'Title': title, 'Genre': ep})
+        HrMagPlayer().play(video, listitem)
```

Comment:
```python
listitem.setProperty("isPlayable", "true")
xbmcplugin.setResolvedUrl(video, True, listitem)
```
Additionally if you already know the stream type you might want to also use `setMimeType` on the listitem to avoid Kodi to perform HEAD requests to find out the mimetype